### PR TITLE
GPLVM tutorial

### PR DESCRIPTION
**Pull request by leachim**
This is some initial code for the GPLVM model. @torfjelde I am trying to get ADVI working for this example, but there is an error popping up to do with logabsdetjac. Do you mind having a look at this, since you seem to have worked on the bijectors package before.

It is working with the NUTS sampler inference, but running the same model with ADVI throws this error.

### CODE REVIEW

File: tutorials/12-gaussian-process-latent-variable-model/12-gplvm-advi.jl [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+  N, D = size(Y)
+  # dimensions of latent space
+  K = ndim
+  noise = 1e-6
+
+  # Priors
+  α ~ MvLogNormal(MvNormal(K, 1.0))
+  σ ~ MvLogNormal(MvNormal(D, 1.0))
+  # use filldist for Zygote compatibility
+  Z ~ filldist(Normal(0., 1.), K, N)
```
Comment (torfjelde) [inline tutorials/12-gaussian-process-latent-variable-model/12-gplvm-advi.jl:42]: The following will fix it:
```suggestion
  Z_vec ~ filldist(Normal(0., 1.), K * N)
  Z = reshape(Z_vec, K, N)
```

It's due to the meanfield approximation constructed not handling higher-dim arrays properly. I have a PR open with a fix for this, but it never got any attention; I'll try to get that PR merged so the above workaround won't be necessary.

Comment (leachim) [reply tutorials/12-gaussian-process-latent-variable-model/12-gplvm-advi.jl:42]: Great, thanks for the quick response!

Comment (leachim) [reply tutorials/12-gaussian-process-latent-variable-model/12-gplvm-advi.jl]: I am running into some sort of race condition/infinity loop now. It's affecting both MCMC inference and ADVI. For ADVI I get an error message, whereas MCMC just seems to run forever. This happens when I uncomment these 2 lines, I added a reproducible example in the latest commit:
#  using Zygote # Tracker supported? check it?
#  Turing.setadbackend(:zygote)
Should I raise an issue for this?

Comment (cpfiffer) [reply tutorials/12-gaussian-process-latent-variable-model/12-gplvm-advi.jl]: I think so -- I've been running into this on another project too, and honestly I'm kind of glad that someone else can replicate this issue outside of my project.

Comment (torfjelde) [reply tutorials/12-gaussian-process-latent-variable-model/12-gplvm-advi.jl]: This is probably also what's timing out all the CIs 😕

Comment (torfjelde) [reply tutorials/12-gaussian-process-latent-variable-model/12-gplvm-advi.jl]: https://github.com/TuringLang/DistributionsAD.jl/issues/180

**Review by rikhuijzer — COMMENTED** [review]
I'll propose a fix for the CI in a bit. It's some issue with the caching.

Could you make the name of the tutorial consistent with the other names? The pattern is
```
tutorials/02-logistic-regression/02_logistic-regression.jmd
[...]
tutorials/08-multinomial-logistic-regression/08_multinomial-logistic-regression.jmd
[...]
```

Also, it would be great if you can hide some assertions in the tutorial. That way, it's easier to check whether the tutorial also is built correctly in the future with different Julia versions or dependency versions.

Finally, maybe it's a good idea to split this PR up into two PRs. That should speed up the review process.

**Comment by rikhuijzer** [comment]
For example, the CI will fail when assertion blocks such as

    ```julia; echo=false
    @assert false       
    ```

fail. Note that this block is hidden from the output. This functionality is tested in https://github.com/TuringLang/TuringTutorials/blob/master/test/build.jl.

**Comment by rikhuijzer** [comment]
@leachim The following error will be fixed if you merge master into your branch:
```
IOError: open("/home/runner/work/TuringTutorials/TuringTutorials/ClonedTuringTutorials/html/12-gaussian-process-latent-variable-model", 0, 0): no such file or directory (ENOENT)
```

**Comment by leachim** [comment]
> @leachim The following error will be fixed if you merge master into your branch:
> 
> ```
> IOError: open("/home/runner/work/TuringTutorials/TuringTutorials/ClonedTuringTutorials/html/12-gaussian-process-latent-variable-model", 0, 0): no such file or directory (ENOENT)
> ```

Many thanks for your help!

I rebased this on master, hope that's okay now. Let me know if I did something wrong :) 

I'll add some assert later on, this is still a bit of work in progress, so bear with me for a little longer. I will let you know once it's ready for review again.

**Comment by rikhuijzer** [comment]
Great :+1: Thanks. I see that the tests in CI are still broken, but that is my bad and not yours. Let me know when the tutorial works for you, then I'll fix the tests.

**Comment by rikhuijzer** [comment]
> I rebased this on master, hope that's okay now. Let me know if I did something wrong :)

Yeah, I'm afraid that something went wrong. PRs which add a new tutorial should only add three files. For example, to add a new tutorial "my-bayesian-tutorial", the PR should only add the files:
```
tutorials/11-my-bayesian-tutorial/11_my-bayesian-tutorial.jmd
tutorials/11-my-bayesian-tutorial/Project.toml
tutorials/11-my-bayesian-tutorial/Manifest.toml
```
and, locally, the build should succeed when doing
```
julia> using TuringTutorials

julia> build("11-my-bayesian-tutorial");

```

I'm aware that this isn't super easy to do and could be easier. I'm working on that, but also have to do my 40 hours a week job and a book that I'm working on, so it will take a while still :slightly_smiling_face:

**Comment by rikhuijzer** [comment]
It's probably easiest to just copy your files into a new and up-to-date branch and open a new PR.

**Comment by leachim** [comment]
Okay, so I rebased this on master and I also squased some commits, so should be pretty much equivalent to opening a new pull request, but we can keep the discussion history.

I am still working on the tutorial. There is two things that might cause issues at the moment
- I have included two files with data (it's a small example that is commonly used in the literature, and so would be nice to have)
- The run time is really long (more than 10hrs at the moment). I think this is similar to the 10: Bayesian differential equations tutorial. 

I'll try to speed it up more, but not sure how much I can reduce runtime. I am also still working on details and fine-tuning things.

**Comment by rikhuijzer** [comment]
> The run time is really long (more than 10hrs at the moment). I think this is similar to the 10: Bayesian differential equations tutorial.

That's very unfortunate yes, because it makes maintaining the tutorial very difficult. If I look at the other tutorials, it would probably mean that it will be outdated/not working in a few months to a year. Can we find a solution for that? Maybe a smaller dataset or simpler example?

**Comment by rikhuijzer** [comment]
From the logs:
```
Error: ArgumentError: Package Turing not found in current path:
 - Run `import Pkg; Pkg.add("Turing")` to install the Turing package.
```
The build will work locally because you have probably added `Turing.jl` to your global environment, but not to the gplvm tutorial environment.

File: tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd [hunk 2]
```diff
@@ -0,0 +1,271 @@
+---
+title: Gaussian Process Latent Variable Model
+permalink: /:collection/:name/
+redirect_from: tutorials/12-gaussian-process-latent-variable-model/
+---
+
+# Gaussian Process Latent Variable Model
+
+In a previous tutorial, we have discussed latent variable models, in particular probabilistic principal
```
Comment (rikhuijzer) [inline tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd:9]: If you write one sentence per line, then this works much better with the diff. Also, the sentences are currently too long for the GitHub interface and for people who use more advanced editors with delete-sentence and similar shortcuts. For example, Visual Studio Code, Vim or Emacs.

File: tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+component analysis (pPCA). Here, we show how we can extend the mapping provided by pPCA to non-linear mappings between input and output. For more details about the Gaussian Process Latent Variable Model (GPLVM), we refer the reader to the [original
+publication](https://jmlr.org/papers/v6/lawrence05a.html) and a [further extension](http://proceedings.mlr.press/v9/titsias10a/titsias10a.pdf).
+
+In short, the GPVLM is a dimensionality reduction technique that allows us to embed a high-dimensional dataset in a lower-dimensional embedding. Importantly, it provides the advantage that the linear mappings from the embedded space can be non-linearised through the use of Gaussian Processes.
+
+Let's start by loading some dependencies.
+```julia
+# Load Turing.
```
Comment (rikhuijzer) [inline tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd:17]:
```suggestion
```

File: tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+
+Random.seed!(1789);
+```
+
+We use a classical data set of synthetic data that is also used in the original publication.
+```julia
+oil_matrix = readdlm("Data.txt", Float64);
+labels = readdlm("DataLabels.txt", Float64);
+labels = mapslices(x -> findmax(x)[2], labels, dims=2);
+# normalize data
```
Comment (rikhuijzer) [inline tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd:34]:
```suggestion

# normalize data
```

File: tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+dt = fit(ZScoreTransform, oil_matrix, dims=2);
+StatsBase.transform!(dt, oil_matrix);
+```
+
+We will start out, by demonstrating the basic similarity between pPCA (see the tutorial on this topic)
```
Comment (rikhuijzer) [inline tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd:39]:
```suggestion
We will start out by demonstrating the basic similarity between pPCA (see the tutorial on this topic)
```

File: tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+  # Dimensionality of the problem.
+  N, D = size(Y)
+  # dimensions of latent space
+  K = ndim
+  noise = 1e-3
+
+  # Priors
+  α ~ MvLogNormal(MvNormal(K, 1.0))
+  σ ~ LogNormal(0.0, 1.0)
+  Z ~ filldist(Normal(0., 1.), K, N)
```
Comment (rikhuijzer) [inline tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd:77]:
```suggestion
  Z ~ filldist(Normal(0.0, 1.0), K, N)
```
More explicit/clear and also consistent with other uses in the tutorial.

File: tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+
+  kernel = kernel_function(α, σ)
+
+  ## DENSE GP
+  gp = GP(kernel)
+  prior = gp(ColVecs(Z), noise)
+
+  for d in 1:D
+    Y[:, d] ~ prior
```
Comment (rikhuijzer) [inline tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd:86]: Spacing inconsistent. Here is a space (`Y[:, d]`) and above it is not (`x[:,d]`). There is no good reason to not stick to one style and use it everywhere. I would suggest using a space behind the comma. See also "Use whitespace to make the code more readable." at [BlueStyle](https://github.com/invenia/BlueStyle).

File: tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+ndim=2
+n_data=40
+```
+
+```julia
+ppca = pPCA(oil_matrix[1:n_data,1:n_features])
+chain_ppca = sample(ppca, NUTS(), 1000)
+```
+```julia
+w = permutedims(reshape(mean(group(chain_ppca, :w))[:,2], (n_features,n_features)))
```
Comment (rikhuijzer) [inline tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd:110]: Can a simple one liner function be defined for these two? The name of the one-liner could also shed light on what these lines do exactly.

Also applies to the 4 or 5 occurrences of this kind of sentence below.

File: tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+```julia
+gplvm_linear = GPLVM(oil_matrix[1:n_data,1:n_features], linear_kernel, n_features)
+
+chain_linear = sample(gplvm_linear, NUTS(), 800)
+z_mean = permutedims(reshape(mean(group(chain_linear, :Z))[:,2], (n_features, n_data)))'
+alpha_mean = mean(group(chain_linear, :α))[:,2]
+```
+
+```julia
+df_gplvm_linear = DataFrame(z_mean', :auto)
```
Comment (rikhuijzer) [inline tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd:139]: Could you add a small clarification for this block?

File: tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+p1 = df_gplvm|>  @vlplot(:point, x=:z1, y=:z2, color="labels:n")
+```
+
+```julia
+p2 = df_gplvm |>  @vlplot(:point, x=:ard1, y=:ard2, color="labels:n")
+```
+
+### Speeding up inference
+
+Gaussian processes tend to be slow, as they naively require.
```
Comment (rikhuijzer) [inline tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd:190]: end of sentence missing

File: tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+  prior = SparseFiniteGP(fobs, finducing)
+
+  for d in 1:D
+    Y[:, d] ~ prior
+  end
+end
+```
+
+```julia
+ndim=2
```
Comment (rikhuijzer) [inline tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd:239]:
```suggestion
ndim = 2
```
Also next two lines

**Comment by rikhuijzer** [comment]
@leachim, you got the "ERROR: LoadError: LoadError: TypeError: in Type{...} expression, expected UnionAll, got Type{Parsers.Options}" error (more info in https://github.com/JuliaData/CSV.jl/pull/862).

To fix it, simply update the CSV dependency to 0.9.

File: tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+
+In short, the GPVLM is a dimensionality reduction technique that allows us to embed a high-dimensional dataset in a lower-dimensional embedding.
+Importantly, it provides the advantage that the linear mappings from the embedded space can be non-linearised through the use of Gaussian Processes.
+
+Let's start by loading some dependencies.
+```julia
+using Turing
+using AbstractGPs, KernelFunctions, Random, Plots
+
+using Distributions, LinearAlgebra
```
Comment (devmotion) [inline tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd:22]: Turing already loads (or rather reexports) Distributions:

```suggestion
using LinearAlgebra
```

File: tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd [hunk 13]
```diff
@@ -0,0 +1,305 @@
...
+@model function pPCA(x, ::Type{TV} = Array{Float64}) where {TV}
+  # Dimensionality of the problem.
+  N, D = size(x)
+  # latent variable z
+  z ~ filldist(Normal(), D, N)
+  # weights/loadings W
+  w ~ filldist(Normal(), D, D)
+  mu = (w * z)'
+  for d in 1:D
+    x[:, d] ~ MvNormal(mu[:, d], 1.)
```
Comment (devmotion) [inline tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd:70]: This constructor is deprecated:

```suggestion
    x[:, d] ~ MvNormal(mu[:, d], I)
```

File: tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd [hunk 14]
```diff
@@ -0,0 +1,305 @@
...
+@model function GPLVM_linear(Y, K=4,::Type{T} = Float64) where {T}
+
+  # Dimensionality of the problem.
+  N, D = size(Y)
+  # K is the dimension of the latent space
+  @assert K <= D
+  noise = 1e-3
+
+  # Priors
+  α ~ MvLogNormal(MvNormal(K, 1.0))
```
Comment (devmotion) [inline tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd:94]:
```suggestion
  α ~ MvLogNormal(MvNormal(zeros(K), I))
```

File: tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd [hunk 15]
```diff
@@ -0,0 +1,305 @@
...
+  Z ~ filldist(Normal(0.0, 1.0), K, N)
+  mu ~ filldist(Normal(0.0, 1.0), N)
```
Comment (devmotion) [inline tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd:96]: Bit shorter:

```suggestion
  Z ~ filldist(Normal(), K, N)
  mu ~ filldist(Normal(), N)
```

File: tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd [hunk 16]
```diff
@@ -0,0 +1,305 @@
...
+@model function GPLVM(Y, K=4,::Type{T} = Float64) where {T}
+
+  # Dimensionality of the problem.
+  N, D = size(Y)
+  # K is the dimension of the latent space
+  @assert K <= D
+  noise = 1e-3
+
+  # Priors
+  α ~ MvLogNormal(MvNormal(K, 1.0))
```
Comment (devmotion) [inline tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd:114]:
```suggestion
  α ~ MvLogNormal(MvNormal(zeros(K), I))
```

File: tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd [hunk 17]
```diff
@@ -0,0 +1,305 @@
...
+  σ ~ LogNormal(0.0, 1.0)
+  Z ~ filldist(Normal(0.0, 1.0), K, N)
+  mu ~ filldist(Normal(0.0, 1.0), N)
```
Comment (devmotion) [inline tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd:117]:
```suggestion
  Z ~ filldist(Normal(), K, N)
  mu ~ filldist(Normal(), N)
```

File: tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd [hunk 18]
```diff
@@ -0,0 +1,305 @@
...
+In short, the GPVLM is a dimensionality reduction technique that allows us to embed a high-dimensional dataset in a lower-dimensional embedding.
+Importantly, it provides the advantage that the linear mappings from the embedded space can be non-linearised through the use of Gaussian Processes.
+
+Let's start by loading some dependencies.
+```julia
+using Turing
+using AbstractGPs, KernelFunctions, Random, Plots
+
+using Distributions, LinearAlgebra
+using VegaLite, DataFrames, StatsPlots, StatsBase
```
Comment (devmotion) [inline tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd:23]: Is there a specific reason for using both StatsPlots and VegaLite? Seems a bit complicated to use two different plotting frameworks in one tutorial.

File: tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd [hunk 19]
```diff
@@ -0,0 +1,305 @@
...
+  # Dimensionality of the problem.
+  N, D = size(Y)
+  # dimension of latent space
+  @assert K <= D
+  # number of inducing points
+  n_inducing = 25
+  noise = 1e-3
+
+  # Priors
+  α ~ MvLogNormal(MvNormal(K, 1.0))
```
Comment (devmotion) [inline tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd:236]:
```suggestion
  α ~ MvLogNormal(MvNormal(zeros(K), I))
```

File: tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd [hunk 20]
```diff
@@ -0,0 +1,305 @@
...
+  σ ~ LogNormal(1.0, 1.0)
+  Z ~ filldist(Normal(0.0, 1.0), K, N)
+  mu ~ filldist(Normal(0.0, 1.0), N)
```
Comment (devmotion) [inline tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd:239]:
```suggestion
  Z ~ filldist(Normal(), K, N)
  mu ~ filldist(Normal(), N)
```

File: tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd (hunk 15 above)
Comment (devmotion) [reply tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd:96]: FYI `filldist(Normal(), N)` just returns a `MvNormal`.

File: tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd [hunk 21]
```diff
@@ -0,0 +1,305 @@
...
+
+  kernel = σ * SqExponentialKernel() ∘ ARDTransform(α)
```
Comment (devmotion) [inline tutorials/12-gaussian-process-latent-variable-model/12_gaussian-process-latent-variable-model.jmd:241]: Use the function from above?

**Comment by leachim** [comment]
Yes, thanks. There is an open compat helper pull request in RDatasets, so once that is merged, I will update the version of CSV. Currently, RDatasets only supports up to 0.8. 
I managed to get the time requirement down to 2hrs or so, so it should be fine to include this now. Once I have the dependencies fixed, and I had another look, I will remove the draft status and then I would ask you to have another look at it. 
Thanks again!

**Pull request merged**